README.md

```diff
@@ -22,6 +22,15 @@ This repository provides rules for building [Rust][rust] projects with [Bazel](h
   </ul>
 </div>
 
+#### WebAssembly
+
+To build a `rust_binary` for wasm32-unknown-unknown add the `--platforms=//rust/platform:wasm` flag.
```

Comment:
This is a great start, but I think that the requirement to set target platform is quite limiting.

WebAssembly is used outside of browsers / JavaScript host environments, and Wasm modules can be loaded by other build targets in the same Bazel workspace, and those build targets must be built using "native" platforms and not the `//rust:platform:wasm` pseudo-platform.

For example, imagine such `BUILD` file:
```
cc_library(
    name = "runtime",
    srcs = "runtime.cc",
    data = [
        ":wasm",
    ],
)

rust_library(
    name = "wasm",
    srcs = ["wasm.rs"],
    edition = "2018",
    crate_type = "cdylib",
    rustc_flags = ["--target=wasm32-unknown-unknown"],
)
```
where Wasm module compiled using `rust_library` must be loaded by the `runtime` written in C/C++.

The ability to override `--target` is something that I have hacked in my local tree, but ideally we would have something like `rust_wasm_library` that uses `--crate-type=cdylib --target=wasm32-unknown-unknown` to compile Wasm modules.

Basically, we need the ability to indicate Wasm target on a per build target basis, and not globally.

What do you think?

Reply:
I think I had a similar problem where I wanted to include the compiled wasm (compiled as wasm32-unknown-unknown) with `include_bytes` and serve it from the binary (compiled as a native Rust binary). We allow it in this PR by using Bazel transitions. See https://github.com/bazelbuild/rules_rust/pull/240/files#diff-9a42a2c125ed40c511198be1665d0faeR346. This allows us to build the entire build with the default target platform (e.g. x86 + Linux) but temporarily transition to `//rust:platform:wasm` when building the wasm code.

The easiest way is to do this in practice is to make a `rust_wasm_bindgen` target which will force the transition for the dependent `rust_library` (and transitively for all its dependencies). I have a note about this here https://github.com/bazelbuild/rules_rust/pull/240/files/7305a12fd276b0c21432d00d4e2ddd8b15115e13#diff-04c6e90faac2675aa89e2176d2eec7d8R31. As I read my note again I realize that perhaps it's a bit unclear since "transition" is a Bazel-specific word. Do you think it would be clearer if I referenced it in this paragraph?

Basically in this paragraph I was trying to explain how you could get the .wasm file if that's the only thing you wanted.

Reply:
Oh, the transitions are great, I didn't realize that you could do it this way.

I think that `rust_wasm_bindgen` is close to what I want, but after reading that paragraph, I simply ignored it, because I don't want anything related to `wasm-bindgen`, I want pure Rust code that's compiled to `wasm32-unknown-unknown`.

As such, I'm suggesting that we would add `rust_wasm_library` that uses the transitions that you've already added for `rust_wasm_bindgen`, i.e.
```
cc_library(
    name = "runtime",
    srcs = "runtime.cc",
    data = [
        ":wasm",
    ],
)

rust_wasm_library(
    name = "wasm",
    deps = ":library"
)

rust_library(
    name = "library",
    srcs = ["library.rs"],
    edition = "2018",
)
```
Alternatively, since you're already doing it for `proc-macro`, I think that we could use `crate_type = "wasm32"` and simply transition to `wasm32-unknown-unknown` in the `rust_library`, i.e.
```
cc_library(
    name = "runtime",
    srcs = "runtime.cc",
    data = [
        ":library",
    ],
)

rust_library(
    name = "library",
    srcs = ["library.rs"],
    crate_type = "wasm32",
    edition = "2018",
)
```
What do you think?

Reply:
Friendly ping before this gets merged.

Reply:
@PiotrSikora

Reply:
I'm definitely against (4), since this feature has nothing to do with wasm-bindgen.

(3) would still require action from the person building the target (vs person writing `BUILD` file), so it's not much different than requiring extra `--platforms=wasm32` flag, IMHO.

I'm fine with either (1) or (2), with a slight preference for (1), since then you could easily add a target that extends existing `rust_library` to be also compiled as Wasm module (i.e. you could have both native and Wasm build targets built at the same time) by simply adding few lines with `rust_wasm_library` (see the example above). You could achieve the same with (2), but then you'd need to duplicate the complete build target, so it would be more error prone, I think.

Also, there is a precedence for `rust_xxx_library` with `rust_proto_library` and `rust_grpc_library`.

It's fine if you want to add it in a separate PR (i.e. I don't want to block this PR from getting merged), but could you rename it to "Add support for wasm-bindgen", since it's not a complete WebAssembly support?

What do you think?